File: python/fi/evals/core/cloud_registry.py
"""
Cloud eval registry — source of truth for eval metadata.

Fetches the full template list from `/sdk/api/v1/get-evals/` once per
(base_url, api_key) tuple and caches the result. Exposes helpers that
turn user-supplied kwargs into the exact key set the backend will accept,
so the SDK never drifts from the backend when evals are added/renamed.

Why this exists
---------------

The old ``templates.py`` hardcoded Pydantic Input models per template
(``OutputOnly``, ``OutputWithContext``, ``OutputWithExpected``, ...). The
Turing revamp renamed/removed/replaced ~50 templates; the hardcoded
schemas drifted and 28 of 57 stopped working. Going forward, the backend
is authoritative — the SDK reads ``required_keys`` and sends only those.
"""
from __future__ import annotations

import logging
import threading
from typing import Any, Dict, Iterable, List, Optional, Set, Tuple

log = logging.getLogger(__name__)


# Module-level cache: {(base_url, api_key_prefix): {eval_name: info_dict}}
_CACHE: Dict[Tuple[str, str], Dict[str, Dict[str, Any]]] = {}
_CACHE_LOCK = threading.Lock()

# Aliases user kwargs can use → canonical backend keys.
# Maps are checked in order; first match wins.
_KEY_ALIASES: Dict[str, Tuple[str, ...]] = {
    # Bidirectional output↔input — some evals only accept one of the two
    # (e.g. prompt_injection wants `input`, toxicity wants `output`).
    # Aliasing only fires when the canonical key isn't already in user_inputs,
    # so we can't accidentally overwrite an explicit user value.
    "output": ("output", "response", "answer", "generated", "input"),
    "input": ("input", "query", "question", "prompt_input", "output"),
    "context": ("context", "contexts"),
    "expected": ("expected", "expected_output", "expected_response", "ground_truth"),
    "expected_value": ("expected_value", "expected_output", "expected_response", "ground_truth"),
    "generated_value": ("generated_value", "output", "response", "answer"),
    "reference": ("reference", "expected_output", "expected_response", "ground_truth"),
    "hypothesis": ("hypothesis", "output", "response"),
    "text": ("text", "output", "content"),
    "conversation": ("conversation", "messages"),
    "prompt": ("prompt", "instructions", "system_prompt"),
    "system_prompt": ("system_prompt", "prompt", "instructions"),
    "image": ("image", "image_url", "input_image_url"),
    "caption": ("caption", "output"),
    "instruction": ("instruction", "prompt"),
    "instructions": ("instructions", "prompt"),
    "images": ("images", "image_urls", "input_image_urls"),
    "input_pdf": ("input_pdf", "pdf"),
    "json_content": ("json_content", "json", "expected_output"),
    "input_audio": ("input_audio", "audio"),
    "audio": ("audio", "input_audio"),
    "generated_audio": ("generated_audio", "audio", "output"),
    "generated_transcript": ("generated_transcript", "transcript", "output"),
}


def _cache_key(base_url: str, api_key: Optional[str]) -> Tuple[str, str]:
    return (base_url.rstrip("/"), (api_key or "")[:12])


def load_registry(
    base_url: str,
    api_key: Optional[str],
    secret_key: Optional[str],
    *,
    force_refresh: bool = False,
) -> Dict[str, Dict[str, Any]]:
    """Fetch and cache the eval template list. Returns {eval_name: info}."""
    key = _cache_key(base_url, api_key)
    if not force_refresh:
        with _CACHE_LOCK:
            cached = _CACHE.get(key)
        if cached is not None:
            return cached

    # Imported lazily to avoid a circular import at module load time.
    import requests

    url = f"{base_url.rstrip('/')}/sdk/api/v1/get-evals/"
    headers = {}
    if api_key:
        headers["X-Api-Key"] = api_key
    if secret_key:
        headers["X-Secret-Key"] = secret_key

    try:
        resp = requests.get(url, headers=headers, timeout=30)
        resp.raise_for_status()
        body = resp.json()
    except Exception as exc:
        log.warning("Failed to load cloud eval registry from %s: %s", url, exc)
        return {}

    items = body.get("result") or []
    by_name: Dict[str, Dict[str, Any]] = {}
    for item in items:
        name = item.get("name")
        if name:
            by_name[name] = item

    with _CACHE_LOCK:
        _CACHE[key] = by_name
    log.debug("Loaded %d cloud eval templates from %s", len(by_name), url)
    return by_name


def get_template_info(
    name: str,
    *,
    base_url: str,
    api_key: Optional[str] = None,
    secret_key: Optional[str] = None,
) -> Optional[Dict[str, Any]]:
    """Return the backend's config block for a given eval, or None."""
    reg = load_registry(base_url, api_key, secret_key)
    return reg.get(name)


def get_required_keys(
    name: str,
    *,
    base_url: str,
    api_key: Optional[str] = None,
    secret_key: Optional[str] = None,
) -> List[str]:
    """List the keys the backend requires for this eval. Empty list if unknown."""
    info = get_template_info(name, base_url=base_url, api_key=api_key, secret_key=secret_key)
    if not info:
        return []
    return list(info.get("config", {}).get("required_keys", []) or [])


def map_inputs_to_backend(
    name: str,
    user_inputs: Dict[str, Any],
    *,
    base_url: str,
    api_key: Optional[str] = None,
    secret_key: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Produce the exact payload the backend expects for ``eval_name``, by:
      1. Looking up the eval's required_keys from the cached registry.
      2. Taking each required key from user_inputs directly if present.
      3. Otherwise resolving via known aliases (e.g. ``output`` → ``response``).
      4. Dropping any keys the backend doesn't accept (the api is strict).

    If the eval isn't in the registry (unknown name, registry load failed),
    falls back to passing user_inputs through unmodified so the backend
    can return its own validation error.
    """
    required = get_required_keys(
        name, base_url=base_url, api_key=api_key, secret_key=secret_key
    )
    if not required:
        return dict(user_inputs)

    mapped: Dict[str, Any] = {}
    for key in required:
        if key in user_inputs:
            mapped[key] = user_inputs[key]
            continue
        for alias in _KEY_ALIASES.get(key, ()):
            if alias in user_inputs and alias != key:
                mapped[key] = user_inputs[alias]
                break
    return mapped


def list_known_names(
    *,
    base_url: str,
    api_key: Optional[str] = None,
    secret_key: Optional[str] = None,
) -> Set[str]:
    """Names of all evals the backend currently has registered."""
    return set(load_registry(base_url, api_key, secret_key).keys())

File: typescript/ai-evaluation/src/core/cloudRegistry.ts
/**
 * Cloud eval registry — source of truth for eval metadata.
 *
 * Mirrors python/fi/evals/core/cloud_registry.py. Fetches the full template
 * list from `/sdk/api/v1/get-evals/` once per (baseUrl, apiKey) tuple and
 * caches the result. Exposes helpers that turn user-supplied kwargs into
 * the exact key set the backend will accept, so the SDK never drifts from
 * the backend when evals are added/renamed.
 */
import axios from "axios";

type RegistryCacheKey = string; // `${baseUrl}|${apiKeyPrefix}`
type RegistryEntry = Record<string, any>;

const _CACHE = new Map<RegistryCacheKey, Record<string, RegistryEntry>>();

/**
 * Aliases user kwargs can use → canonical backend keys.
 * Direct key match in user_inputs always wins over an alias lookup.
 */
const KEY_ALIASES: Record<string, readonly string[]> = {
    // Bidirectional output↔input — some evals only accept one (e.g.
    // prompt_injection wants `input`, toxicity wants `output`).
    output: ["output", "response", "answer", "generated", "input"],
    input: ["input", "query", "question", "prompt_input", "output"],
    context: ["context", "contexts"],
    expected: ["expected", "expected_output", "expected_response", "ground_truth"],
    expected_value: ["expected_value", "expected_output", "expected_response", "ground_truth"],
    generated_value: ["generated_value", "output", "response", "answer"],
    reference: ["reference", "expected_output", "expected_response", "ground_truth"],
    hypothesis: ["hypothesis", "output", "response"],
    text: ["text", "output", "content"],
    conversation: ["conversation", "messages"],
    prompt: ["prompt", "instructions", "system_prompt"],
    system_prompt: ["system_prompt", "prompt", "instructions"],
    image: ["image", "image_url", "input_image_url"],
    caption: ["caption", "output"],
    instruction: ["instruction", "prompt"],
    instructions: ["instructions", "prompt"],
    images: ["images", "image_urls", "input_image_urls"],
    input_pdf: ["input_pdf", "pdf"],
    json_content: ["json_content", "json", "expected_output"],
    input_audio: ["input_audio", "audio"],
    audio: ["audio", "input_audio"],
    generated_audio: ["generated_audio", "audio", "output"],
    generated_transcript: ["generated_transcript", "transcript", "output"],
};


function cacheKey(baseUrl: string, apiKey?: string): RegistryCacheKey {
    return `${baseUrl.replace(/\/$/, "")}|${(apiKey || "").slice(0, 12)}`;
}

export interface LoadRegistryOptions {
    baseUrl: string;
    apiKey?: string;
    secretKey?: string;
    forceRefresh?: boolean;
}

export async function loadRegistry(
    opts: LoadRegistryOptions
): Promise<Record<string, RegistryEntry>> {
    const key = cacheKey(opts.baseUrl, opts.apiKey);
    if (!opts.forceRefresh && _CACHE.has(key)) {
        return _CACHE.get(key)!;
    }

    const url = `${opts.baseUrl.replace(/\/$/, "")}/sdk/api/v1/get-evals/`;
    const headers: Record<string, string> = {};
    if (opts.apiKey) headers["X-Api-Key"] = opts.apiKey;
    if (opts.secretKey) headers["X-Secret-Key"] = opts.secretKey;

    try {
        const resp = await axios.get(url, { headers, timeout: 30000 });
        const items: any[] = resp.data?.result || [];
        const byName: Record<string, RegistryEntry> = {};
        for (const item of items) {
            if (item?.name) byName[item.name] = item;
        }
        _CACHE.set(key, byName);
        return byName;
    } catch (err) {
        // Soft-fail — let the caller fall back to pass-through so the
        // api can surface its own validation error.
        console.warn(`Failed to load cloud eval registry from ${url}:`, err);
        return {};
    }
}


export async function getTemplateInfo(
    name: string,
    opts: LoadRegistryOptions
): Promise<RegistryEntry | undefined> {
    const reg = await loadRegistry(opts);
    return reg[name];
}


export async function getRequiredKeys(
    name: string,
    opts: LoadRegistryOptions
): Promise<string[]> {
    const info = await getTemplateInfo(name, opts);
    return info?.config?.required_keys ?? [];
}


/**
 * Build the exact payload the backend expects for an eval, by:
 *   1. Looking up the eval's required_keys from the cached registry.
 *   2. Taking each required key from userInputs directly if present.
 *   3. Otherwise resolving via known aliases.
 *   4. Dropping any keys the backend doesn't accept (the api is strict).
 *
 * If the eval isn't in the registry (unknown name, registry load failed),
 * falls back to passing userInputs through unmodified.
 */
export async function mapInputsToBackend(
    name: string,
    userInputs: Record<string, any>,
    opts: LoadRegistryOptions
): Promise<Record<string, any>> {
    const required = await getRequiredKeys(name, opts);
    if (required.length === 0) {
        return { ...userInputs };
    }

    const mapped: Record<string, any> = {};
    for (const key of required) {
        if (key in userInputs) {
            mapped[key] = userInputs[key];
            continue;
        }
        const aliases = KEY_ALIASES[key] || [];
        for (const alias of aliases) {
            if (alias !== key && alias in userInputs) {
                mapped[key] = userInputs[alias];
                break;
            }
        }
    }
    return mapped;
}


/** Synchronous variant — uses cached data. Throws if not preloaded. */
export function mapInputsToBackendSync(
    name: string,
    userInputs: Record<string, any>,
    baseUrl: string,
    apiKey?: string
): Record<string, any> {
    const cached = _CACHE.get(cacheKey(baseUrl, apiKey));
    if (!cached || !cached[name]) {
        return { ...userInputs };
    }
    const required: string[] = cached[name]?.config?.required_keys ?? [];
    if (!required.length) {
        return { ...userInputs };
    }
    const mapped: Record<string, any> = {};
    for (const key of required) {
        if (key in userInputs) {
            mapped[key] = userInputs[key];
            continue;
        }
        const aliases = KEY_ALIASES[key] || [];
        for (const alias of aliases) {
            if (alias !== key && alias in userInputs) {
                mapped[key] = userInputs[alias];
                break;
            }
        }
    }
    return mapped;
}


/** Clear cache (exposed for tests). */
export function __clearRegistryCache(): void {
    _CACHE.clear();
}
